heft-plugins/heft-webpack4-plugin/src/WebpackPlugin.ts

```diff
@@ -1,157 +1,186 @@
+  IWebpackVersions
 } from './shared';
 import { WebpackConfigurationLoader } from './WebpackConfigurationLoader';
 
-const webpack: typeof import('webpack') = Import.lazy('webpack', require);
-
-const PLUGIN_NAME: string = 'WebpackPlugin';
+/**
+ * @public
+ */
+export const PLUGIN_NAME: string = 'WebpackPlugin';
 const WEBPACK_DEV_SERVER_PACKAGE_NAME: string = 'webpack-dev-server';
 const WEBPACK_DEV_SERVER_ENV_VAR_NAME: string = 'WEBPACK_DEV_SERVER';
 
-interface IWebpackVersions {
-  webpackVersion: string;
-  webpackDevServerVersion: string;
-}
-
 /**
  * @internal
  */
-export class WebpackPlugin implements IHeftPlugin {
-  public readonly pluginName: string = PLUGIN_NAME;
+export default class WebpackPlugin implements IHeftTaskPlugin {
+  private _webpack: typeof TWebpack | undefined;
+  private _webpackVersions: IWebpackVersions | undefined;
+  private _loadedWebpackConfiguration: IWebpackConfiguration | null | undefined;
+
+  public readonly accessor: IWebpackPluginAccessor = {
+    onEmitWebpackVersionsHook: new AsyncParallelHook(['webpackVersions']),
+    onConfigureWebpackHook: new AsyncSeriesWaterfallHook(['webpackConfiguration']),
+    onAfterConfigureWebpackHook: new AsyncParallelHook(['webpackConfiguration']),
+    onEmitStatsHook: new AsyncParallelHook(['webpackStats'])
+  };
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    this.accessor.onConfigureWebpackHook!.tapPromise(
+      PLUGIN_NAME,
+      async (existingConfiguration: IWebpackConfiguration | null) => {
+        if (existingConfiguration) {
+          taskSession.logger.terminal.writeVerboseLine(
+            'Skipping loading webpack config file because the webpack config has already been set.'
+          );
+          return existingConfiguration;
+        } else {
+          const configurationLoader: WebpackConfigurationLoader = new WebpackConfigurationLoader(
+            taskSession.logger,
+            production,
+            serveMode
+          );
+          return await configurationLoader.tryLoadWebpackConfigAsync(heftConfiguration.buildFolder);
+        }
+      }
+    );
+
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Obtain the finalized webpack configuration
+      const webpackConfiguration: IWebpackConfiguration | null = await this._getWebpackConfigurationAsync();
+      if (webpackConfiguration) {
+        const webpackConfigurationArray: IWebpackConfigurationWithDevServer[] = Array.isArray(
+          webpackConfiguration
+        )
+          ? webpackConfiguration
+          : [webpackConfiguration];
+
+        // Add each output path to the clean list
+        for (const config of webpackConfigurationArray) {
+          if (config.output?.path) {
```

Comment:
Funny story, you can actually walk out of the output path when emitting files in webpack. Do we currently have the means to provide additional static clean paths for a given task?

Reply:
We can provide additional static paths to clean on the phase, yes (phase was chosen since phases own execution of cleaning for tasks), so that would be done in heft.json.